File: src/python/lib/Doctor.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


from dataclasses import dataclass
from typing import ClassVar
from .worker import Worker


@dataclass
class Doctor(Worker):
    """Help on the Doctor module:

    Name

        Doctor 

    Description 

        The doctor class is developed to simulate a person that is a doctor.

        This class inherits attributes and methods from the Worker class.

        - property decorator: getters for class. 
        - setter decorator: setters for class.

    Args

        first_name: first name of doctor. 

        last_name: last name of doctor.

        gender: gender of doctor.

        height: height of doctor. 

        weight: weight of doctor.

        age: age of doctor. 

        weekly hours: weekly hours of doctor.

        salary: salary of doctor.

        type: type of doctor.

    Usage
        
        doctor = Doctor("Ale", "Pagan", "M", "5'8", 170, 22, 40, 90000, "Generalist")

    Package Contents

        - _first_name
        - _last_name
        - _gender
        - _height
        - _age
        - _weekly_hours
        - _salary
        - _person_count 
        - _worker_count
        - _doctor_count
        - talk()
        - first_name() || first_name(first_name:str)
        - last_name() || last_name(last_name:str)
        - gender() || gender(gender:str)
        - weight() || weight(weight:int)
        - age() || age(age:int)
        - weekly_hours() || weekly_hours(weekly_hours:int)
        - salary() || salary(salary:int)
        - type() || type(type:str)

    Author 

        Osvaldo Aquino
    """

    _type: str # represents the type of doctor.
    _doctor_count : ClassVar[int] = 0 # static variable used to keep track of each doctor that is created.


    def __post_init__(self):
        """ Post init function used to increment the global class variables used in this object."""

        Doctor._person_count+=1
        Doctor._worker_count+=1
        Doctor._doctor_count+=1


    @property
    def type(self) -> str:
        """ Returns

                type of doctor.

            Usage

                doctor.type """

        return self._type


    @type.setter
    def type(self, type: str) -> None:
        """ Sets the type attribute of the doctor given a string.

            Args

                type(str): Represents the new type that will be set.

            Usage

                doctor.type = "Cardiologist" """

        self._type = type


    def talk(self) -> str:
        """ Prints out useful information about the doctor object.

            Usage
            
                doctor.talk()"""

        print(f'Hello! My name is {self._first_name} {self._last_name}, I am a {self._type} doctor that works {self.weekly_hours} a week and I have a salary of {self._salary}.')

File: src/cpp/lib/Doctor.cpp
/**
 * @file person.hpp
 * @author Osvaldo Aquino (osvaldo.aquino@upr.edu)
 * @brief TODO: Document - Abstract class for person collection.
 * @version 0.1
 * @date 2022-01-10
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include <lib/headers/doctor.hpp>


Doctor::Doctor(std::string fname, std::string lname, std::string g, std::string h, int w, int age, int hours, int salary, std::string type)
{
    this->_first_name = fname;
    this-> _last_name = lname;
    this->_gender = g;
    this->_height = h;
    this->_weight = w;
    this->_age = age;
    this->_weekly_hours = hours;
    this->_type = type;
    _person_count++;
    _worker_count++;
    _doctor_count++;
}


void Doctor::talk(void)
{
    std::cout << "Hello! My name is" << this->_first_name << this->_last_name << ", I am a" << this->_type << "doctor that works" 
    << this->_weekly_hours << "a week and I have a salary of" << this->_salary << "." << std::endl;;
}
